OpenRA.Mods.Common/Traits/AutoTarget.cs

```diff
@@ -167,43 +164,57 @@ void Attack(Actor self, Actor targetActor, bool allowMove)
 
 		Actor ChooseTarget(Actor self, WDist range, bool allowMove)
 		{
-			var inRange = self.World.FindActorsInCircle(self.CenterPosition, range)
-				.Where(a =>
-					!a.TraitsImplementing<IPreventsAutoTarget>().Any(t => t.PreventsAutoTarget(a, self)));
+			var actorsByArmament = new Dictionary<Armament, List<Actor>>();
+			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, range);
+			foreach (var actor in actorsInRange)
+			{
+				if (PreventsAutoTarget(self, actor) || !self.Owner.CanTargetActor(actor))
+					continue;
+
+				// Select only the first compatible armament for each actor: if this actor is selected
+				// it will be thanks to the first armament anyways, since that is the first selection
+				// criterion
+				var target = Target.FromActor(actor);
+				var armaments = attack.ChooseArmamentsForTarget(target, false);
+				if (!allowMove)
+					armaments = armaments.Where(arm =>
+						target.IsInRange(self.CenterPosition, arm.MaxRange()) &&
+						!target.IsInRange(self.CenterPosition, arm.Weapon.MinRange));
+
+				var armament = armaments.FirstOrDefault();
+				if (armament == null)
+					continue;
+
+				List<Actor> actors;
+				if (actorsByArmament.TryGetValue(armament, out actors))
+					actors.Add(actor);
```

Comment:
Somehow I don't get the logic here... Why are we adding stuff to a field that isn't used anymore after?
Should this be `actorsByArmament` instead of `actors`?